### DISCUSSION TITLE
NPM 7 life cycle script changes

### DISCUSSION THREAD

**Discussion by dmcghan** [npm/feedback#592] [CLI]
I have an npm module (not yet published to npm) that makes use of the `postinstall` script to install dependencies that are external to npm. In npm v7, some behaviors related to running this script changed and I'm trying to figure out what to do about the changes.

The first change is that the scripts are now run as background scripts. The [doc here](https://docs.npmjs.com/cli/v7/using-npm/scripts#life-cycle-scripts) for the life cycle scripts seems to document this change somewhat under **Prepare**, where it says: 

> As of npm@7 these scripts run in the background. To see the output, run with: --foreground-scripts.

However, I've observed this change affecting not only 'prepare', but 'install' as well (perhaps this is a documentation bug).

The problem with running the scripts in the background is that `console.log` output isn't shown, so the user can't see the progress as the external dependencies are installed. Also, if an error occurs, the user doesn't know about it. I see there's a documented workaround (using --foreground-scripts) that does preserve the previous behavior (at least with `npm run postinstall` - I've not tested `install` alone yet).

However, it's a very "non-standard" practice to ask users to run the scripts with that flag - people want to just do `npm install foo`. Is there an alternative way to get the previous behavior back without using the flag? Perhaps a setting in the package.json?

The second change, and this is related to the first change, is the order of operations. In previous versions of npm, `postinstall` would run after 'install' and I was able to take for granted that the dependencies listed in the package.json would be there and I could use them in the `postinstall` script. However, in the npm v7, the `postinstall` script often errors out complaining that a given package can't be resolved. The package that it errors out on seems to be somewhat random. If I install the package manually and then rerun the `postinstall` script again with `npm run postinstall` then it succeeds just fine.

Does anyone know what may be happening here and how I could fix it?

**Comment by ljharb** [comment]
> Is there an alternative way to get the previous behavior back without using the flag? Perhaps a setting in the package.json?

There isn't any other way to get that behavior back; it was used by some package authors to spam users.

**Reply by dmcghan** [reply]
I see, thanks! Do you have any thoughts on the "timing" issue? For example, `postinstall` recently errored out with: 
`npm ERR! Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'tar' imported from /Users/doe/project/bin/install-other-dependencies.mjs`

The 'tar' module is listed in the package.json, so I would think when `postinstall` runs, 'tar' would be ready to go.

The reason I ask is that, in the lucky case that no errors occur, then at least the script would finish installing external dependencies. To increase the chances of success, I'd like to figure out why this isn't working as expected.

Thanks again for your help!

**Reply by dmcghan** [reply]
Nevermind. I'm now planning on making the install process a two-step process. The first step will be to run the `npm install` and the second step will be to run something like `npm run install-deps`. This should address both the timing and the feedback issues.

Thanks again.
